### PR TITLE
Don't fire warnings/errors about "partial availability" on Apple platforms

### PR DESCRIPTION
**Pull request by rcgordon**
In reference to https://github.com/curl/curl/issues/1333 ...

This will fix the compiler warnings (that manifest as compiler errors) when building curl for macOS and targeting older OSes with -mmacosx-version-min=10.?.

This makes sure that all the newer symbols are marked weak, compared against NULL at runtime, with a minimum of fuss.

Note that the configure script changes just turn off the compiler warning for the clock_gettime() tests, but it might be smarter to make sure the script doesn't add -Werror=partial-availability until the very end of the script, so no other tests in the future fail because of it. That change is safe to leave in here in any case, though.

### CODE REVIEW

**Comment by nickzman** [comment]
Question... Why do we even check for partial availability in the first place? I would be in favor of removing that error entirely.

**Comment by rcgordon** [comment]
oh wow, the CI builds alert if a line is longer than 79 characters, that's pretty hardcore.  :)   I'll fix that.

**Comment by rcgordon** [comment]
> Question... Why do we even check for partial availability in the first place? I would be in favor of removing that error entirely.

I'm not against removing it, but it has been there awhile, and I believe it caught a binary compatibility bug with the use of clock_gettime(), which has been available on Linux forever, but just arrived in macOS for the latest version, 10.2.

But if you want to remove it instead, I approve.  :)

**Comment by markovicpp** [comment]
If the warning/error/check has been introduced just for the clock_gettime() then I think this is an overkill. Redeclaring correctly declared functions is way too much code to maintain.

**Comment by rcgordon** [comment]
> If the warning/error/check has been introduced just for the clock_gettime() then I think this is an overkill. Redeclaring correctly declared functions is way too much code to maintain.

Fwiw, I suppose the flag was added because there was worry about things like the clock_gettime problem happening in the future: https://github.com/curl/curl/issues/1069

But again, just removing the flag is fine with me; this patch is just working within the existing parameters, but I agree the flag causes more problems than it solves. I won't be offended if this is closed without being merged if the flag gets removed instead.

**Comment by markovicpp** [comment]
Thanks for clarification, I see now that the problem was caused by sudden change of availability of function symbol in new SDK which afterwards triggered configure script condition.

AFAICS it’s up to decision if a feature check should be performed during compile time vs run time. libcurl favours compile time which is easier to maintain across multiple platforms, even though it means that some features cannot be used when running on system version supporting it.

**Comment by markovicpp** [comment]
And I just verified that the `-Werror=partial-availability` is used in configuration phase to determine if the required range of system versions support the function. If the configuration test fails during the compilation thanks to that error it implies that the function is not supported across specified system range.

If that error is to be removed, parts of your patch with the runtime check for `clock_gettime` symbol will still be needed to avoid crash and providing fallback functionality.

Alternatively the `-Werror=partial-availability` could be used only in the configuration phase.

**Comment by bagder** [comment]
I will appreciate feedback and comments from people on/with mac in regards to this issue and this PR. It is beyond me. I can't merge until I get more data to go on.

**Comment by rcgordon** [comment]
Fwiw, my preference is that the `-Werror=partial-availability` flag just be removed outright, and this patch was working on the assumption that working to keep that flag was worth the effort, but I don't believe that's a true assumption at this point. Even in the configuration phase, the flag would result in building a less functional curl if older systems are targeted, even though the code in question will already do the right thing on both old and new systems.

My take is that the flag will maybe catch a future bug, but not much faster than the appropriate Mac user will file a bug report and/or PR here about it anyhow.

I'll defer to what @markovicpp and others think, but my personal recommendation is: close this PR unmerged and take a different PR that just removes the flag.

**Comment by markovicpp** [comment]
I agree with @rcgordon.

The clash of different concepts for providing backwards compatibility is inevitable across multiple operating systems and while it is possible to stick with just configuration checks, this really limits curl to the machine/system it has been built on and prohibits replacement of deprecated methods, which is quite common in Apple/Darwin world. However this comes with a higher price of maintenance and understanding how different systems work.

I also vote that this PR is closed without merging, and that the new patch should be created addressing removal of that particular compiler flag simultaneously with adding runtime check for clock_gettime which was the reason why the flag has been added.

Basically that original fix would be reverted (along with setting minimal system version deployment) and replaced with runtime check for `clock_gettime` symbol as long as it is supported outside Darwin OS line:

`  if(&clock_gettime && 0 == clock_gettime(CLOCK_MONOTONIC, &tsnow)) {`

Btw. the same code in slight variations exists in curl source base 3 times.
